File: scripts/convert.py
#!/usr/bin/env python3
# Copyright (C) 2026 Anas Khan
# SPDX-License-Identifier: GPL-3.0-or-later
"""Render Mermaid diagrams as sketch-style PNG or SVG images."""

from __future__ import annotations

import argparse
import os
import re
import shutil
import subprocess
import sys
from pathlib import Path

SKILL_ROOT = Path(__file__).resolve().parent.parent

FENCE_RE = re.compile(r"```mermaid[^\n]*\n(.*?)```", re.S)
KIND_RE = re.compile(r"^(sequenceDiagram|stateDiagram(?:-v2)?|classDiagram|erDiagram|gantt|pie|gitGraph|flowchart|graph)\b", re.I)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Convert Mermaid fences or a .mmd file into images."
    )
    parser.add_argument("markdown", nargs="?", help="Markdown file with mermaid fences")
    parser.add_argument("--input", "-i", help="Single Mermaid file, or - for stdin")
    parser.add_argument("--output", "-o", help="Output image for a single diagram")
    parser.add_argument("--out-dir", help="Directory for markdown images (default: images/)")
    parser.add_argument("--prefix", default="diagram", help="Filename prefix")
    parser.add_argument("--bg", default="white", help="white, transparent, or #hex")
    parser.add_argument("--font-size", type=int, default=None, dest="font_size")
    parser.add_argument("--font-color", default=None, dest="font_color")
    parser.add_argument("--format", choices=("png", "svg"), default="png")
    parser.add_argument(
        "--replace",
        action="store_true",
        help="Rewrite mermaid fences to image tags in the markdown file",
    )
    return parser.parse_args()


def extract_blocks(markdown: str) -> list[str]:
    return [match.group(1).strip() for match in FENCE_RE.finditer(markdown)]


def diagram_kind(definition: str) -> str:
    for line in definition.splitlines():
        text = line.strip()
        if not text or text.startswith("%%") or text.startswith("---"):
            continue
        match = KIND_RE.match(text)
        if match:
            token = match.group(1).lower()
            if token.startswith("sequence"):
                return "sequence"
            if token.startswith("state"):
                return "state"
            if token.startswith("class"):
                return "class"
            if token.startswith("er"):
                return "er"
            if token.startswith("flowchart") or token.startswith("graph"):
                return "flowchart"
            return token
        return "diagram"
    return "diagram"


def ensure_renderer() -> None:
    if (SKILL_ROOT / "node_modules" / "playwright").exists():
        return
    npm = shutil.which("npm")
    if not npm:
        raise SystemExit("Node.js and npm are required. Install them, then run npm install in the skill folder.")
    subprocess.run([npm, "install"], cwd=SKILL_ROOT, check=True)


def render(definition: str, output: Path, *, fmt: str, background: str, font_size: int | None, font_color: str | None) -> None:
    ensure_renderer()
    node = shutil.which("node")
    if not node:
        raise SystemExit("Node.js is required to render Excalidraw diagrams.")
    command = [
        node,
        str(SKILL_ROOT / "scripts" / "render.mjs"),
        "--input",
        "-",
        "--output",
        str(output),
        "--format",
        fmt,
        "--bg",
        background,
    ]
    if font_size:
        command.extend(["--font-size", str(font_size)])
    if font_color:
        command.extend(["--font-color", font_color])
    result = subprocess.run(
        command,
        input=definition.encode("utf-8"),
        cwd=SKILL_ROOT,
        capture_output=True,
    )
    if result.returncode != 0:
        detail = (result.stderr or result.stdout).decode("utf-8", "replace").strip()
        raise SystemExit(detail or "Renderer failed")


def convert_markdown(args: argparse.Namespace) -> None:
    markdown_path = Path(args.markdown).expanduser().resolve()
    markdown = markdown_path.read_text()
    blocks = extract_blocks(markdown)
    if not blocks:
        raise SystemExit(f"No mermaid fences in {markdown_path}")

    out_dir = Path(args.out_dir).expanduser().resolve() if args.out_dir else markdown_path.parent / "images"
    counts: dict[str, int] = {}
    replacements: list[tuple[str, str]] = []

    for definition in blocks:
        kind = diagram_kind(definition)
        counts[kind] = counts.get(kind, 0) + 1
        filename = f"{args.prefix}-{kind}-{counts[kind]}.{args.format}"
        output = out_dir / filename
        render(
            definition,
            output,
            fmt=args.format,
            background=args.bg,
            font_size=args.font_size,
            font_color=args.font_color,
        )
        print(output)
        rel = os.path.relpath(output, markdown_path.parent).replace(os.sep, "/")
        replacements.append((definition, rel))

    if args.replace:
        cursor = 0

        def swap(match: re.Match[str]) -> str:
            nonlocal cursor
            definition, rel = replacements[cursor]
            cursor += 1
            return f"<!-- mermaid\n{definition}\n-->\n\n![{diagram_kind(definition)} diagram]({rel})"

        markdown_path.write_text(FENCE_RE.sub(swap, markdown))
        print(markdown_path)


def convert_single(args: argparse.Namespace) -> None:
    if args.input == "-" or (not args.input and not sys.stdin.isatty()):
        definition = sys.stdin.read()
    elif args.input:
        definition = Path(args.input).expanduser().read_text()
    else:
        raise SystemExit("Provide a markdown file, --input, or stdin.")

    definition = definition.strip()
    if not definition:
        raise SystemExit("Mermaid input is empty")

    output = Path(args.output).expanduser() if args.output else Path(f"diagram.{args.format}")
    render(
        definition,
        output,
        fmt=args.format,
        background=args.bg,
        font_size=args.font_size,
        font_color=args.font_color,
    )
    print(output)


def main() -> None:
    args = parse_args()
    if args.markdown:
        convert_markdown(args)
    else:
        convert_single(args)


if __name__ == "__main__":
    main()

File: scripts/convert.mjs
#!/usr/bin/env node

import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, extname, join, relative, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { HELP, parseArgs } from "./lib/args.mjs";
import { createConverter } from "./lib/browser.mjs";
import {
  diagramKind,
  extractMermaidBlocks,
  replaceMermaidBlocks,
} from "./lib/markdown.mjs";

const readStdin = async () => {
  const chunks = [];
  for await (const chunk of process.stdin) {
    chunks.push(chunk);
  }
  return Buffer.concat(chunks).toString("utf8");
};

const normalizeBackground = (value) => {
  if (!value || value === "white") {
    return "#ffffff";
  }
  if (value === "transparent") {
    return "transparent";
  }
  return value;
};

const inferFormat = (output, format) => {
  if (format) {
    return format;
  }
  const extension = extname(output || "").toLowerCase();
  if (extension === ".svg") return "svg";
  if (extension === ".excalidraw" || extension === ".json") return "excalidraw";
  return "png";
};

const writeResult = (output, format, result) => {
  mkdirSync(dirname(output), { recursive: true });
  if (format === "svg") {
    writeFileSync(output, result.svg);
    return;
  }
  if (format === "excalidraw") {
    writeFileSync(output, `${JSON.stringify(result.scene, null, 2)}\n`);
    return;
  }
  writeFileSync(output, Buffer.from(result.pngBase64, "base64"));
};

const convertOne = async (converter, definition, options, output) => {
  const format = inferFormat(output, options.format);
  const result = await converter.convert(definition, {
    format,
    background: normalizeBackground(options.background),
    fontSize: options.fontSize,
    fontColor: options.fontColor,
    scale: options.scale,
    padding: options.padding,
  });
  writeResult(output, format, result);
  return { output, format, elementCount: result.elementCount };
};

const runMarkdown = async (converter, options) => {
  const markdownPath = resolve(options.markdown);
  const markdown = readFileSync(markdownPath, "utf8");
  const blocks = extractMermaidBlocks(markdown);
  if (blocks.length === 0) {
    throw new Error(`No mermaid fences found in ${markdownPath}`);
  }

  const outDir = resolve(options.outDir || join(dirname(markdownPath), "images"));
  mkdirSync(outDir, { recursive: true });
  const format = options.format || "png";
  const extension =
    format === "svg" ? "svg" : format === "excalidraw" ? "excalidraw" : "png";

  const replacements = [];
  const counts = {};
  for (const block of blocks) {
    const kind = diagramKind(block.definition);
    counts[kind] = (counts[kind] || 0) + 1;
    const filename = `${options.prefix}-${kind}-${counts[kind]}.${extension}`;
    const output = join(outDir, filename);
    const converted = await convertOne(converter, block.definition, options, output);
    const src = relative(dirname(markdownPath), output).split("\\").join("/");
    replacements.push({
      alt: `${kind} diagram`,
      src,
      definition: block.definition,
      ...converted,
    });
    console.log(`wrote ${output} (${converted.elementCount} elements)`);
  }

  if (options.replace) {
    const next = replaceMermaidBlocks(markdown, replacements, {
      keepSource: options.keepSource,
    });
    writeFileSync(markdownPath, next);
    console.log(`updated ${markdownPath}`);
  }

  return replacements;
};

const main = async () => {
  const options = parseArgs(process.argv.slice(2));
  if (options.help) {
    process.stdout.write(HELP);
    return;
  }

  if (Number.isNaN(options.fontSize) || options.fontSize <= 0) {
    throw new Error("--font-size must be a positive number");
  }
  if (Number.isNaN(options.scale) || options.scale <= 0) {
    throw new Error("--scale must be a positive number");
  }

  const converter = await createConverter();
  try {
    if (options.markdown) {
      await runMarkdown(converter, options);
      return;
    }

    let definition = "";
    if (options.input === "-" || (!options.input && !process.stdin.isTTY)) {
      definition = await readStdin();
    } else if (options.input) {
      definition = readFileSync(resolve(options.input), "utf8");
    } else {
      throw new Error("Provide --input, stdin, or --markdown. See --help.");
    }

    definition = definition.trim();
    if (!definition) {
      throw new Error("Mermaid input is empty");
    }

    const output = resolve(options.output || `diagram.${options.format || "png"}`);
    const converted = await convertOne(converter, definition, options, output);
    console.log(`wrote ${converted.output} (${converted.elementCount} elements)`);
  } finally {
    await converter.close();
  }
};

const isMain = process.argv[1] && fileURLToPath(import.meta.url) === resolve(process.argv[1]);
if (isMain) {
  main().catch((error) => {
    console.error(error.message || error);
    process.exitCode = 1;
  });
}
